### PR TITLE
retouch : fix shift compute if distortion before

### LINKED ISSUE #8675: Spots and retouch stamps misaligned if distortion changes
**Describe the bug/issue**

darktable draws masks in RAW coordinates by belief masks should be anchored to pixels and move with them if they are to be distorted at a later time but lower in the pipe. Spots and Retouch stamps show a gross misalignment when the image is deformed. Liquify is ok though.

**To Reproduce**

Use
grid

Put clone stamps at corners with lens correction, crop and rotate and perspective correction modules off.
Screenshot_20210414_125241

Enable perspective module with arbitrary deformation: mismatch
Screenshot_20210414_125305

Enable crop and rotate with arbitrary keystone deformation : mismatch
Screenshot_20210414_125350

Test with liquify, no distortion on:
Screenshot_20210414_125552

Apply distortions : match with both perspective module and crop and rotate keystone
Screenshot_20210414_125726

Screenshot_20210414_125740

**Expected behavior**

Same as liquify.

### PR DESCRIPTION
**Pull request by AlicVB**
partially fix #8675

this is a first draft, to fix some spot/retouch issues if placed after a distort module. This "just" fix the shift computation. but at least, the source center is at the right place now :)

Note, this is NOT the definitive fix, just a POC : 
- printf all around
- retouch not fixed
- no legacy params

EDIT : no more POC. ready to review, test and merged :)

### CODE REVIEW

**Comment by AlicVB** [comment]
Here is a clean version, ready to test and review.

reminder of the issue : retouch and spots don't use the right source position if a module that distort the image is placed before in the pipe (this is the case in default(3.0) pipe order.

1. I've only done the work for retouch.
2. old forms created are untouched via an upgrade of the params. only new forms are correct. this is to ensure backward compatibility.

Finally, I've not done the same for spots removal : 
- I really think that it should be deprecated, retouch cover the same things, with a better algo by default, and it don't need more click or whatever...
- spot has a problem in how forms and their algos id are stored in params (only the first form do it right) This is not a problem for now, but it would had lot of headhaches if we want to do the same as in retouch...

**Comment by elstoc** [comment]
Looks good. I think it might need a redraw if another module changes. If I have retouch masks shown and toggle lens correction the image distorts but the on-screen shapes stay in the same location (relative to the screen).

lens correction off:
Screenshot_2021-04-16_22-24-08

lens correction enabled while retouch module expanded:
Screenshot_2021-04-16_22-24-33

After closing and reopening retouch module (to force shape redraw):
Screenshot_2021-04-16_22-25-29

**Comment by elstoc** [comment]
> I really think that [spots] should be deprecated, retouch cover the same things

Maybe we could let the user choose the default retouch algorithm so that they can choose "clone" by default if they mostly prefer "spots-like" behaviour?

**Comment by AlicVB** [comment]
@elstoc : about mask redraw : it's not related to this, but it's global for masks (I would even say for all drawings on canvas). Another thing to fix :)



> Maybe we could let the user choose the default retouch algorithm so that they can choose "clone" by default if they mostly prefer "spots-like" behaviour?

maybe... I would better do something like "remember the last algo used" to avoid cluttering the ui with another option...

**Comment by elstoc** [comment]
> I would better do something like "remember the last algo used"

I would prefer it to be predictable from one image to the next. If we want to avoid cluttering up I'd prefer either leaving it as a hidden (darktablerc only) setting, or just leave the default as heal.

**Comment by AlicVB** [comment]
> I would prefer it to be predictable from one image to the next. If we want to avoid cluttering up I'd prefer either leaving it as a hidden (darktablerc only) setting, or just leave the default as heal.

agree. For me, heal as default is quite sane, but I can imagine that it's not the case for everyone !
maybe it's a good candidate for a pref inside the hamburger button, like we do for other module iirc ? (but is it worth the pain ?)

**Comment by elstoc** [comment]
Ctrl+click is better for me to set default - we can amend the tooltip so it's not too hidden

**Comment by AlicVB** [comment]
with last commit, if you shift+click on algo button, it becomes the default one (for next images).
- I've not used ctrl+click, as it is already used to force change the algo of the current form (I've updated the tooltip accordingly)
- this new feature is not described in any tooltip, due to how `dt_iop_togglebutton_new` work (it automagically construct the tooltip for modifier key, but just for ctrl)... I'm not sure how to change that without doing crappy workaround or make a lot of change all around

**Comment by elstoc** [comment]
> I've not used ctrl+click

Fine with me

> this new feature is not described in any tooltip, due to how dt_iop_togglebutton_new work

tbh I'd just override the tooltip immediately after the widgets are created. The alternative would be to amend dt_iop_togglebutton_new to include a shift label but that might be overkill for this one usage. I think the other stuff is used primarily for keyboard shortcuts and we probably don't need a shortcut for these.

> for the bug about masks redraw, I prefer to let that apart and fix it in another PR

Agreed

**Comment by AlicVB** [comment]
@elstoc : I've forced pushed last commit with right tooltips...

**Comment by elstoc** [comment]
Tooltips look good. 

Do we really want to change the default *and* change the current tool with SHIFT or just change the default? I was surprised when it did both. Perhaps a toast should appear advising that the default has been changed?

Also, if you change the default and then discard history (select original then compress), it goes back to the previous default, unless you exit the darkroom and re-enter.

**Comment by AlicVB** [comment]
> Do we really want to change the default _and_ change the current tool with SHIFT or just change the default? I was surprised when it did both

hmm... I can't reproduce, (and I'm surprised that it can happen, looking at the code...) : to be sure : 
1. you add a retouch form with "heal" selected
2. you shit-click on clone
3. => this change the algo used form your form to "clone" ???

for the rest/discard, I'll have a look...
Thanks again for your in-dep testing !

**Comment by elstoc** [comment]
> => this change the algo used form your form to "clone"

it toggles the button but I honestly didn't test if it actually changed the algo

**Comment by AlicVB** [comment]
new version with toast msg, and discard history or reset module correctly handled (thanks to point me in the right direction)


> it toggles the button but I honestly didn't test if it actually changed the algo

sorry, I've misunderstood what you was meaning... I've reread your comment, and yes, it change the default + switch to this tool... I don't find that surprising for my part... (and I've added the toast message to be more clear)

**Comment by elstoc** [comment]
Thanks @AlicVB all good for me now.

> I don't find that surprising for my part...

Not a massive issue for me. I'm fine to leave it as is.

**Comment by elstoc** [comment]
Maybe change the OP if this really _is_ the definitive fix now?

**Comment by AlicVB** [comment]
@elstoc : yes this is definitive and ready to merge for my part. (not sure to understand 'OP')
@TurboGit : I think that it should be for 3.6, too, as it's a bug fix quite severe in the actual default pipe order (3.0)

**Comment by elstoc** [comment]
> (not sure to understand 'OP')

Original Post (also can mean original poster). It's a good idea to keep the first post of a PR up-to-date with its current status so that people don't have to read the whole conversation when browsing. Also useful for when I come along to do user manual updates. The OP should essentially contain the basic documentation of what the PR does and how it works.

**Comment by TurboGit** [comment]
@AlicVB : agreed, I have scheduled this for 3.6.

**Comment by elstoc** [comment]
@TurboGit with the ability to set the default algorithm I think spots could probably be deprecated now (assuming that retouch/clone uses the same algorithm)

File: src/iop/retouch.c [hunk 1]
```diff
@@ -1698,6 +1787,21 @@ static gboolean rt_select_algorithm_callback(GtkToggleButton *togglebutton, GdkE
 
   dt_dev_add_history_item(darktable.develop, self, TRUE);
 
+  // if we have the shift key pressed, we set it as default
+  if(dt_modifier_is(e->state, GDK_SHIFT_MASK))
+  {
+    dt_conf_set_int("plugins/darkroom/retouch/default_algo", p->algorithm);
+    // and we show a toat msg to confirm
+    if(p->algorithm == DT_IOP_RETOUCH_CLONE)
+      dt_toast_log(N_("default tool changed to %s"), N_("cloning"));
```
Comment (TurboGit) [inline src/iop/retouch.c:1796]: This should be dt_control_log and not dt_toast_log. Same for all occurrences. The toast log is for very ephemeral information like changing opacity and is display top-center. The control log is where such information should be sent to be consistent.

**Review by TurboGit — CHANGES_REQUESTED** [review]
Minor changes, otherwise code ok and all works fine. Another very nice piece of work! Thanks.

File: src/iop/retouch.c (hunk 1 above)
Comment (elstoc) [reply src/iop/retouch.c:1796]: On reflection I agree. If your attention is on the right-hand panel (clicking buttons), dt_toast_log can easily be missed.

Comment (AlicVB) [reply src/iop/retouch.c]: agree... and done :)

**Review by TurboGit — APPROVED** [review]
All good then, thanks!

**Pull request merged**